### ISSUE TITLE
skorch.NeuralNet auto score method given criterion

### ISSUE THREAD

**Issue by hsorsky** [skorch-dev/skorch#628]
For ``torch.nn.Module``s passed to ``skorch.NeuralNet``, I think it would make sense to automatically use the passed criterion in the score function if no score is implemented. E.g.

```python
class MyModule(torch.nn.Module):
    def __init__(self, ...):
        ...

class MyLoss(_Loss):
    def __init__(self, ...):
        ...

net = NeuralNet(
    module=MyModule,
    ...
    criterion=MyLoss,
)

net.fit(X, y) # fits by minimising MyLoss's forward method
net.score(X, y) # scores by evaluating MyLoss's forward method
```

**Comment by BenjaminBossan** [comment]
That's an interesting suggestion, since it's not super straightforward to achieve this result at the moment. I would probably not go so far as to make this the default behavior, since it could have some backwards-incompatible consequences (for instance, [`check_scoring`](https://github.com/scikit-learn/scikit-learn/blob/02309ffbdaae45af75b0e87946aab9aedb6b3634/sklearn/metrics/_scorer.py#L420) checks if an estimator has the `score` method, which our nets don't).

A possibility I could see is through a mixin class. The idea would be that I can opt into this by defining:

```python
class MyNet(NeuralNet, ScoringMixin):
    """Gives NeuralNet a score method"""
```

This mixin could reside in the skorch helpers.

Another option would be through some kind of argument (`net = NeuralNet(..., auto_score=True)`) but I don't like adding methods dynamically if there are other, less "magical" ways.

@hsorsky Do you want to work on this?

**Comment by asberk** [comment]
@BenjaminBossan Could you explain what you mean re backwards-incompatible consequences? 

I'm new to using `skorch`. My team is adopting it, so I'm trying to make the transition from `ignite`. So far, I'm confused that the easiest "skorchy" way to get the loss on a dataset `(X, y)` seems to one of:
```python
# this
net.criterion_(nnbc.evaluation_step(X), skorch.utils.to_tensor(y, torch.device('cpu'))).numpy()
# or this
nnbc.get_loss(nnbc.forward(X), torch.tensor(y).float()).numpy()
```
Am I missing something? This seems like too basic a functionality not to have a shorter expression for (especially given that `net.predict` already exists). 

Anyway, if such functionality does not already exist, would you be able to describe in slightly more detail how the change should be implemented so as to avoid the troubles you alluded to above? I would be happy to work on this issue.

**Comment by BenjaminBossan** [comment]
I thought some more about this and I think a better solution is to create a scoring function with a signature like this:

```python
def loss_scoring(net, X, y=None):
    ...
```

Then it's easy to implement your own net that supports the `score` method:

```python
class MyNet(NeuralNet):
    def score(self, X, y=None):
        return loss_scoring(self, X, y)
```

I think it's a bit easier to use than the mixin and it's also easy to pass around to `GridSearchCV` etc. without having to subclass `NeuralNet`.

**Comment by asberk** [comment]
Hi @BenjaminBossan 
This sounds fairly straightforward. As before, I could probably use some pointers with best practices…
Right now, I've got some working code (with some simple tests passing), but which needs to be adapted to how skorch "does things". 

```python
def loss_scoring(net, X, y=None):
    from skorch.dataset import unpack_data

    net.check_is_fitted()
    accumulator = Accumulator(net.criterion_.reduction)
    dataset = net.get_dataset(X, y)
    iterator = net.get_iterator(dataset, training=False)

    for data in iterator:
        Xi, yi = unpack_data(data)
        yp = net.evaluation_step(Xi, training=False)
        loss = net.criterion_(yp, yi)
        accumulator(loss.item(), Xi.size(0))
    loss_value = accumulator.accumulate()
    return loss_value
```
Right now I have an `Accumulator` class (definition not included) which plays the role of tallying up the losses from each batch. This sounds like something that `PassThroughScoring` does, but I haven't had time to look to see if it will play nicely right out of the box. What would you recommend for this purpose? 

Side question: in which module would you house `loss_scoring`? I've had it in `skorch.utils`, but I could see it belonging elsewhere.

**Comment by BenjaminBossan** [comment]
@asberk Thanks a lot for working on this. I will gladly give some guidance if needed. For that, I would propose that you open a (draft) pull request with your current work, including tests. It will make it much easier for me to judge the overall implementation and to give hints from there.

> Side question: in which module would you house `loss_scoring`? I've had it in `skorch.utils`, but I could see it belonging elsewhere.

It depends. Let's assume that this function will require access to `PassThroughScoring`, then we cannot put it into `utils.py` because of possible circular dependencies. In that case, it might even be okay to create a new module for this function, e.g. `scoring.py`.

**Comment by BenjaminBossan** [comment]
@asberk Any news?

**Comment by BenjaminBossan** [comment]
@hsorsky You can now find this feature on the master branch via #665.
